### DISCUSSION TITLE
Can xonsh help solve the "spaces in interpreter" problem on Unix?

### DISCUSSION THREAD

**Discussion by jaraco** [xonsh/xonsh#5315] [Q&A] [answered]
According to [Wikipedia Shebang (Unix)](https://en.wikipedia.org/wiki/Shebang_(Unix)#cite_note-linux-8), the reference specification for the shebang is in the [execve man page](https://linux.die.net/man/2/execve), which is erily silent on the specification of the _interpreter_ portion.

> The `interpreter` must be a valid pathname for an executable which is not itself a script.

It can be readily demonstrated that [Unix doesn't support spaces in filenames](https://blog.jaraco.com/unix-doesnt-support-spaces-in-filenames/) because some (most) flavors of Unix will choke if there's a space in the interpreter name of the shebang. This causes [problems](https://github.com/pypa/pipx/issues/1198) when systems like pipx try to honor platform guidelines. And although Linux users might think they're immune, they're not.

```xsh
docker run -it jaraco/multipy-tox bash

mkdir /foo\ bar
py -m venv /foo\ bar/env
/foo\ bar/env/bin/pip install -q docutils
/foo\ bar/env/bin/rst2html.py
# bash: /foo bar/env/bin/rst2html.py: /foo: bad interpreter: No such file or directory
```

This constraint almost certainly bullies Linux users into avoiding spaces in path names ever.

Today, I learned that xonsh is already leading the way by supporting quoted executables:

```xsh
xonsh
echo @("""#!"/Users/jaraco/Library/Application Support/pipx/venvs/mercurial/bin/python"
print("hello")""") > hello

chmod u+x hello
./hello
# hello      # It works in xonsh

# It is not, however, supported by bash nor zsh:
bash -c './hello'
# bash: line 1: ./hello: cannot execute: required file not found
zsh -c './hello'
# zsh:1: ./hello: bad interpreter: "/Users/jaraco/Library/Application: no such file or directory
```

Presumably that works because of the use of shlex here:

https://github.com/xonsh/xonsh/blob/04cba39a53ccb7bb8e8a5a23939ed8d000687d6b/xonsh/procs/specs.py#L123

This syntax is also supported in the Python Launcher that's distributed with Windows.

Has anyone on the xonsh team worked with maintainers of other shells? How about Linux itself or Unix?
Any advice on where one might start to fix this long-standing limitation?

**Comment by anki-code** [comment]
If it's already working in xonsh. What do you want to fix?

**Reply by jaraco** [reply]
I'm wondering if the xonsh team has any established channels of communication with other shell maintainers or awareness of forums where this issue might be discussed. Is there any instinct on whether this problem could/should be tackled in Unix/Linux first or whether other shells might be amenable to offering support independent of `execve`. I know this isn't the right forum, but I imagine the xonsh maintainers may have attended conferences or have met with other shell maintainers and have some insight into how best to approach this issue.

**Reply by anki-code** [reply, marked as answer]
I think the answer is no. Xonsh was made with Windows in mind. But other old shells had no windows support.
